profiler/src/ProfilerEngine/Datadog.Profiler.Native/LiveObjectsProvider.cpp

```diff
@@ -100,30 +99,62 @@ void LiveObjectsProvider::OnGarbageCollectionEnd(
     });
 }
 
-std::list<std::shared_ptr<Sample>> LiveObjectsProvider::GetSamples()
+class LiveObjectsEnumerator : public SamplesEnumerator
 {
-    // return the live object samples
-    std::list<std::shared_ptr<Sample>> liveObjectsSamples;
+public:
+    LiveObjectsEnumerator(std::size_t size) :
+        _currentPos{0}
+    {
+        _samples.reserve(size);
+    }
 
-    // limit lock scope
+    void Add(std::shared_ptr<Sample> sample)
     {
-        std::lock_guard<std::mutex> lock(_liveObjectsLock);
+        _samples.push_back(std::move(sample));
+    }
 
-        int64_t currentTimestamp = OpSysTools::GetHighPrecisionTimestamp();
-        for (auto const& info : _monitoredObjects)
-        {
-            // gen2 objects are candidates for leaking however collections could be rare and only gen1 objects
-            // are available for live heap profiling
-            auto sample = info.GetSample();
-            liveObjectsSamples.push_back(sample);
-
-            // update samples lifetime
-            sample->ReplaceLabel(Label{Sample::ObjectLifetimeLabel, std::to_string(sample->GetTimeStamp() - currentTimestamp)});
-            sample->ReplaceLabel(Label{Sample::ObjectGenerationLabel, info.IsGen2() ? Gen2 : Gen1});
-        }
+    // Inherited via SamplesEnumerator
+    std::size_t size() const override
+    {
+        return _samples.size();
+    }
+
+    bool MoveNext(std::shared_ptr<Sample>& sample) override
+    {
+        if (_currentPos >= _samples.size())
+            return false;
+
+        sample = _samples[_currentPos++];
+        return true;
+    }
+
+    std::vector<std::shared_ptr<Sample>> _samples;
+    std::size_t _currentPos;
+};
+
+std::unique_ptr<SamplesEnumerator> LiveObjectsProvider::GetSamples()
+{
+    std::lock_guard<std::mutex> lock(_liveObjectsLock);
+
+    int64_t currentTimestamp = OpSysTools::GetHighPrecisionTimestamp();
+    std::size_t nbSamples = 0;
+
+    // OPTIM maybe use an allocator
+    auto samples = std::make_unique<LiveObjectsEnumerator>(_monitoredObjects.size());
+
+    for (auto const& info : _monitoredObjects)
+    {
+        // gen2 objects are candidates for leaking however collections could be rare and only gen1 objects
+        // are available for live heap profiling
+        auto sample = info.GetSample();
+        samples->Add(sample);
```

Comment:
to avoid confusion, move this line after the sample updates

Reply:
✔️